js/battle.js

```diff
@@ -1802,6 +1804,41 @@ var Side = (function () {
 		pokemon.statbarElem = this.battle.statElem.children().last();
 		this.updateStatbar(pokemon, true);
 		pokemon.side.updateSidebar();
+
+		if (pokemon.replacedIllusion && !this.battle.hackedPokemon) {
+			if (this.pokemon.indexOf(pokemon) < 0) {
```

Comment:
As far as I'm aware, this will never be true?

Reply:
This is true if a Pokemon with Illusion faints without revealing itself to be an Illusion.

From what I can tell something like the following happens:
- `Side#newPokemon` creates the new pokemon that switches in and adds it to the team.
- Finds a duplicate on the team, and removes the new Pokemon that it just added to fix this.
- Returns the new Pokemon it just removed from the team.

Reply:
If it finds a duplicate on the team, it should remove the old Pokemon, not the new Pokemon it just added...

Reply:
That's how `newPokemon` works right now though with regards to Illusion.

https://github.com/Zarel/Pokemon-Showdown-Client/blob/master/js/battle.js#L1747 In every case I can figure out, this splices out the last duplicate it finds, which for `pokemon.length > maxPokemon` will always be the new Pokemon it just pushed onto the team.

Reply:
Oh, that's entirely unintended. I wonder if that's the cause of a number of our bugs. It should definitely be splicing out the old copy.

Reply:
Double checking before I make any other changes, but isn't this logic also reversed here https://github.com/Zarel/Pokemon-Showdown-Client/blob/master/js/battle.js#L1745? If `poke1` is fainted but `poke2` isn't, why is `poke2` the one that gets spliced out and the fainted one remains?

Reply:
That's... probably a mistake.

Reply:
Wait, no, that's right.

Reply:
The number of fainted Pokémon should be kept the same. We assume that the fainted Pokémon is the one that's not an Illlusion, and if we're wrong, we mark a Zoroark as fainted or something.